# Data Preparation - *Alfendio Alif Faudisyah*

## Cleaning Data

Unduh data CSV dari: https://github.com/realpython/python-data-cleaning/blob/master/Datasets/BL-Flickr-Images-Book.csv

Import library yang dibutuhkan

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("BL-Flickr-Images-Book.csv")

In [3]:
print(df.dtypes)

Identifier                  int64
Edition Statement          object
Place of Publication       object
Date of Publication        object
Publisher                  object
Title                      object
Author                     object
Contributors               object
Corporate Author          float64
Corporate Contributors    float64
Former owner               object
Engraver                  float64
Issuance type              object
Flickr URL                 object
Shelfmarks                 object
dtype: object


lihat statsistik data

In [4]:
df.describe()

Identifier  Corporate Author  Corporate Contributors  Engraver
count  8.287000e+03               0.0                     0.0       0.0
mean   2.017344e+06               NaN                     NaN       NaN
std    1.190379e+06               NaN                     NaN       NaN
min    2.060000e+02               NaN                     NaN       NaN
25%    9.157875e+05               NaN                     NaN       NaN
50%    2.043707e+06               NaN                     NaN       NaN
75%    3.047430e+06               NaN                     NaN       NaN
max    4.160339e+06               NaN                     NaN       NaN

In [5]:
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

Buang fungsi yang tidak diinginkan

In [6]:
yg_dihapus = ['Edition Statement','Corporate Author'

In [7]:
df.drop(yg_dihapus, inplace=True, axis=1)

In [8]:
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                  Contributors  \
0      A. A.                               FORBES, Walter.   
1  A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   
2  A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   
3  A., E. S.                   Appleyard, Ernest Silvanus.   
4  A., E. S.                           BROOME, John Henry.   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [9]:
yg_dihapus = ['Corporate Contributors','Former owner','Engraver','Issuance type','Flickr URL','Shelfmarks']

In [10]:
df.drop(yg_dihapus, inplace=True, axis=1)

In [11]:
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                  Contributors  
0      A. A.                               FORBES, Walter.  
1  A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness  
2  A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness  
3  A., E. S.                   Appleyard, Ernest Silvanus.  
4  A., E. S.                           BROOME, John Henry.

Setting identifier menjadi indeks

2. Mengubah indeks di data

In [12]:
df['Identifier'].is_unique

True

In [13]:
df = df.set_index('Identifier')

In [14]:
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                            Contributors  
Identifier                                                
206                                      FORBES, Walter.  
216         BLAZE DE BURY, Marie Pauline Rose - Baroness  
218         BLAZE DE BURY, Marie Pauline Rose - Baroness  
472                          Appleyard, Ernest Silvanus.  
480                                  BROOME, John Henry.

Akses berdasarkan label indeks

In [15]:
df.loc[206]

Place of Publication                               London
Date of Publication                           1879 [1878]
Publisher                                S. Tinsley & Co.
Title                   Walter Forbes. [A novel.] By A. A
Author                                              A. A.
Contributors                              FORBES, Walter.
Name: 206, dtype: object

In [16]:
df.loc[472]

Place of Publication                                               London
Date of Publication                                                  1851
Publisher                                                   James Darling
Title                   Welsh Sketches, chiefly ecclesiastical, to the...
Author                                                          A., E. S.
Contributors                                  Appleyard, Ernest Silvanus.
Name: 472, dtype: object

Rapikan data

3. Merapikan fields 

In [17]:
df.dtypes.value_counts()

object    6
dtype: int64

In [18]:
df.loc[1868:,'Date of Publication'].head()

Identifier
1905           1888
1929    1839, 38-54
2836           1897
2854           1865
2956        1860-63
Name: Date of Publication, dtype: object

In [19]:
df.loc[1905:,'Date of Publication'].head()

Identifier
1905           1888
1929    1839, 38-54
2836           1897
2854           1865
2956        1860-63
Name: Date of Publication, dtype: object

In [20]:
df.loc[1905:,'Date of Publication'].head(10)

Identifier
1905           1888
1929    1839, 38-54
2836           1897
2854           1865
2956        1860-63
2957           1873
3017           1866
3131           1899
4598           1814
4884           1820
Name: Date of Publication, dtype: object

In [21]:
df.loc[1905:,'Date of Publication'].head(15)

Identifier
1905                  1888
1929           1839, 38-54
2836                  1897
2854                  1865
2956               1860-63
2957                  1873
3017                  1866
3131                  1899
4598                  1814
4884                  1820
4976                  1800
5382    1847, 48 [1846-48]
5385               [1897?]
5389               [1897?]
5432                  1893
Name: Date of Publication, dtype: object

- hilangkan tanggal lain dalam kurung siku
- konversi rentang tanggal
- hilangkan tanggal yang tidak jelas [1897?] --> Nan
- konversi Nan

Butuh regex (regular expression)

\d mewakili sembarang digit
{4} berapa digit yang mau diambil
'^ disesuaikan dengan awal string

In [23]:
regex = r'^(\d{4})'

In [25]:
ekstrak = df['Date of Publication'].str.extract(r'^(\d{4})', expand=False)
ekstrak.head()

Identifier
206    1879
216    1868
218    1869
472    1851
480    1857
Name: Date of Publication, dtype: object

In [26]:
ekstrak = df['Date of Publication'].str.extract(r'^(\d{4})', expand=False)
ekstrak.head(15)

Identifier
206     1879
216     1868
218     1869
472     1851
480     1857
481     1875
519     1872
667      NaN
874     1676
1143    1679
1280    1802
1808    1859
1905    1888
1929    1839
2836    1897
Name: Date of Publication, dtype: object

type masih object, ubah ke numeric

In [28]:
df['Date of Publication'] = pd.to_numeric(ekstrak)
df['Date of Publication'].dtype

dtype('float64')

In [ ]:
4. Membersihkan kolom

In [29]:
df['Place of Publication'].head(15)

Identifier
206                                  London
216                London; Virtue & Yorston
218                                  London
472                                  London
480                                  London
481                                  London
519                                  London
667     pp. 40. G. Bryan & Co: Oxford, 1898
874                                 London]
1143                                 London
1280                               Coventry
1808                            Christiania
1905                                Firenze
1929                              Amsterdam
2836                                 Savona
Name: Place of Publication, dtype: object

In [30]:
df.loc[667]

Place of Publication                  pp. 40. G. Bryan & Co: Oxford, 1898
Date of Publication                                                   NaN
Publisher                                                             NaN
Title                   The Coming of Spring, and other poems. By J. A...
Author                                                      A., J.|A., J.
Contributors                                ANDREWS, J. - Writer of Verse
Name: 667, dtype: object

In [31]:
df.loc[4157862]

Place of Publication                                  Newcastle-upon-Tyne
Date of Publication                                                1867.0
Publisher                                                      T. Fordyce
Title                   Local Records; or, Historical Register of rema...
Author                      FORDYCE, T. - Printer, of Newcastle-upon-Tyne
Contributors             SYKES, John - Bookseller, of Newcastle-upon-Tyne
Name: 4157862, dtype: object

In [32]:
df.loc[472]

Place of Publication                                               London
Date of Publication                                                1851.0
Publisher                                                   James Darling
Title                   Welsh Sketches, chiefly ecclesiastical, to the...
Author                                                          A., E. S.
Contributors                                  Appleyard, Ernest Silvanus.
Name: 472, dtype: object

In [33]:
df['Place of Publication'].tail(15)

Identifier
4115138    G. Eld: London, 1608
4116063                  London
4117526                 Leipzig
4117583                  London
4117749                New York
4117751                New York
4117752                  London
4156359                New York
4157746                  London
4157862     Newcastle-upon-Tyne
4158088                  London
4158128                   Derby
4159563                  London
4159587     Newcastle upon Tyne
4160339                  London
Name: Place of Publication, dtype: object

In [34]:
df.loc[4115138]

Place of Publication                                 G. Eld: London, 1608
Date of Publication                                                   NaN
Publisher                                                             NaN
Title                   The Revenger's Tagaedie. As it hath beene sund...
Author                                                                NaN
Contributors                                             TOURNEUR, Cyril.
Name: 4115138, dtype: object

contains melihat apa yang mengandung
replace ganti

In [36]:
publikasi = df['Place of Publication']
#london kecil ubah ke London besar
london = publikasi.str.contains('London')
london[:5]

Identifier
206    True
216    True
218    True
472    True
480    True
Name: Place of Publication, dtype: bool

In [37]:
oxford = publikasi.str.contains('Oxford')

biar langsung tidak satu-satu bisa pake fungsinya numpy yaitu np.where

In [38]:
df['Place of Publication'] = np.where(london, 'London',
                                     np.where(oxford, 'Oxford',
                                             publikasi.str.replace('-', ' '))) #tanda - ganti dengan spasi

In [39]:
df['Place of Publication'].head()

Identifier
206    London
216    London
218    London
472    London
480    London
Name: Place of Publication, dtype: object

In [40]:
df['Place of Publication'].head(15)

Identifier
206          London
216          London
218          London
472          London
480          London
481          London
519          London
667          Oxford
874          London
1143         London
1280       Coventry
1808    Christiania
1905        Firenze
1929      Amsterdam
2836         Savona
Name: Place of Publication, dtype: object

In [41]:
df['Place of Publication'].tail(15)

Identifier
4115138                 London
4116063                 London
4117526                Leipzig
4117583                 London
4117749               New York
4117751               New York
4117752                 London
4156359               New York
4157746                 London
4157862    Newcastle upon Tyne
4158088                 London
4158128                  Derby
4159563                 London
4159587    Newcastle upon Tyne
4160339                 London
Name: Place of Publication, dtype: object

In [42]:
df.head()

Place of Publication  Date of Publication              Publisher  \
Identifier                                                                    
206                      London               1879.0       S. Tinsley & Co.   
216                      London               1868.0           Virtue & Co.   
218                      London               1869.0  Bradbury, Evans & Co.   
472                      London               1851.0          James Darling   
480                      London               1857.0   Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                            Contributors  
Identifier                                                
206                                      FORBES, Walter.  
216         BLAZE DE BURY, Marie Pauline Rose - Baroness  
218         BLAZE DE BURY, Marie Pauline Rose - Baroness  
472                          Appleyard, Ernest Silvanus.  
480                                  BROOME, John Henry.

Kita gunakan dataset yang lain/baru, sumber data dari: https://github.com/realpython/python-data-cleaning/blob/master/Datasets/university_towns.txt

5. Membersihkan dataset dengan Applymap

In [44]:
university_town = []
with open("university_towns.txt") as file:
    for line in file: #baca perbaris
        if '[edit]' in line:
            state = line
        else:
            university_town.append((state,line))
university_town[:5]

[('Alabama[edit]\n', 'Auburn (Auburn University)[1]\n'),
 ('Alabama[edit]\n', 'Florence (University of North Alabama)\n'),
 ('Alabama[edit]\n', 'Jacksonville (Jacksonville State University)[2]\n'),
 ('Alabama[edit]\n', 'Livingston (University of West Alabama)[2]\n'),
 ('Alabama[edit]\n', 'Montevallo (University of Montevallo)[2]\n')]

Bungkus/wrap data yang ada, mana yg state mana yang region name

In [46]:
df_kota = pd.DataFrame(university_town, columns=['State','RegionName'])
df_kota.head()

State                                         RegionName
0  Alabama[edit]\n                    Auburn (Auburn University)[1]\n
1  Alabama[edit]\n           Florence (University of North Alabama)\n
2  Alabama[edit]\n  Jacksonville (Jacksonville State University)[2]\n
3  Alabama[edit]\n       Livingston (University of West Alabama)[2]\n
4  Alabama[edit]\n         Montevallo (University of Montevallo)[2]\n

In [47]:
df_kota = pd.DataFrame(university_town, columns=['State','RegionName'])
df_kota.head(15)

State                                         RegionName
0    Alabama[edit]\n                    Auburn (Auburn University)[1]\n
1    Alabama[edit]\n           Florence (University of North Alabama)\n
2    Alabama[edit]\n  Jacksonville (Jacksonville State University)[2]\n
3    Alabama[edit]\n       Livingston (University of West Alabama)[2]\n
4    Alabama[edit]\n         Montevallo (University of Montevallo)[2]\n
5    Alabama[edit]\n                        Troy (Troy University)[2]\n
6    Alabama[edit]\n  Tuscaloosa (University of Alabama, Stillman Co...
7    Alabama[edit]\n                Tuskegee (Tuskegee University)[5]\n
8     Alaska[edit]\n    Fairbanks (University of Alaska Fairbanks)[2]\n
9    Arizona[edit]\n       Flagstaff (Northern Arizona University)[6]\n
10   Arizona[edit]\n                 Tempe (Arizona State University)\n
11   Arizona[edit]\n                   Tucson (University of Arizona)\n
12  Arkansas[edit]\n  Arkadelphia (Henderson State University, Ouach...
13  Arkansas[edit]\n  Conway (Central Baptist College, Hendrix Colle...
14  Arkansas[edit]\n         Fayetteville (University of Arkansas)[7]\n

Coba Applymap, hanya satu kali
Applymap untuk memetakan semua di data frame

In [ ]:
#contoh_data = {0 : ['mock','python','real','numpy']}
#               1 : ['Dataset','pandas','python','clean']

Buat class dulu baru Applymap

In [51]:
def get_citystate(item):
    if '(' in item:
        return item[:item.find('(')]
    elif '[' in item:
        return item[:item.find('[')]
    else:
        return item

Semua data di df_kota pakai format citystate

In [52]:
df_kota = df_kota.applymap(get_citystate)

In [53]:
df_kota.head()

State     RegionName
0  Alabama        Auburn 
1  Alabama      Florence 
2  Alabama  Jacksonville 
3  Alabama    Livingston 
4  Alabama    Montevallo

In [54]:
df_kota.head(15)

State     RegionName
0    Alabama        Auburn 
1    Alabama      Florence 
2    Alabama  Jacksonville 
3    Alabama    Livingston 
4    Alabama    Montevallo 
5    Alabama          Troy 
6    Alabama    Tuscaloosa 
7    Alabama      Tuskegee 
8     Alaska     Fairbanks 
9    Arizona     Flagstaff 
10   Arizona         Tempe 
11   Arizona        Tucson 
12  Arkansas   Arkadelphia 
13  Arkansas        Conway 
14  Arkansas  Fayetteville

#### Alfendio Alif Faudisyah - 672019222 - Metodologi Data Science B - Pertemuan 7 - 2 Maret 2022